src/pdu.c

```diff
@@ -703,52 +737,125 @@ coap_pdu_parse_opt(coap_pdu_t *pdu) {
-      case COAP_OPTION_SIZE1:         if (len > 4) goto bad;              break;
-      case COAP_OPTION_NORESPONSE:    if (len > 1) goto bad;              break;
-      default:
-        ;
+      else if (!coap_pdu_parse_opt_base(pdu, len)) {
+        goto bad;
       }
       continue;
 bad:
-      coap_log(LOG_DEBUG,
-               "coap_pdu_parse: %d.%02d: option %u has bad length %u\n",
-               pdu->code >> 5, pdu->code & 0x1F, pdu->max_opt, len);
+      coap_log(LOG_WARNING,
+            "coap_pdu_parse: %d.%02d: offset %u option %u has bad length %u\n",
+               pdu->code >> 5, pdu->code & 0x1F,
+               (int)(opt_last - pdu->token - pdu->token_length), pdu->max_opt,
+               len);
       good = 0;
     }
 
+    if (!good) {
+      /*
+       * Dump the options in the PDU for analysis, space separated except
+       * error options which are prefixed by *
+       * Two rows - hex and ascii (if printable)
+       */
+      static char outbuf[100];
+      uint32_t tlen;
+      int outbuflen;
+
+      opt = pdu->token + pdu->token_length;
+      length = pdu->used_size - pdu->token_length;
+      pdu->max_opt = 0;
+
+      snprintf(outbuf, sizeof(outbuf), "O: ");
+      while (length > 0 && *opt != COAP_PAYLOAD_START) {
+        coap_opt_t *opt_last = opt;
+        size_t optsize = next_option_safe(&opt, &length, &pdu->max_opt);
+        const uint32_t len =
+          optsize ? coap_opt_length((const uint8_t *)opt - optsize) : 0;
+        if (!optsize) {
+          opt = pdu->token + pdu->token_length +
+                pdu->used_size - pdu->token_length;
+          length = 0;
+          goto badp;
+        }
+        if (COAP_PDU_IS_SIGNALING(pdu)) {
+          if (!coap_pdu_parse_opt_csm(pdu, len))
+            goto badp;
+        }
+        else if (!coap_pdu_parse_opt_base(pdu, len)) {
+          goto badp;
+        }
+        tlen = opt - opt_last;
+        outbuflen = strlen(outbuf);
+        snprintf(&outbuf[outbuflen], sizeof(outbuf)-outbuflen, " ");
+        while (tlen--) {
+          outbuflen = strlen(outbuf);
+          snprintf(&outbuf[outbuflen], sizeof(outbuf)-outbuflen,
+                   "%02x", *opt_last++);
+        }
+        continue;
+badp:
+        tlen = opt - opt_last;
+        outbuflen = strlen(outbuf);
+        snprintf(&outbuf[outbuflen], sizeof(outbuf)-outbuflen, "*");
+        while (tlen--) {
+          outbuflen = strlen(outbuf);
+          snprintf(&outbuf[outbuflen], sizeof(outbuf)-outbuflen,
+                   "%02x", *opt_last++);
+        }
+      }
+      if (length && *opt == COAP_PAYLOAD_START) {
+        outbuflen = strlen(outbuf);
+        snprintf(&outbuf[outbuflen], sizeof(outbuf)-outbuflen,
+                   " %02x", *opt);
+      }
+      coap_log(LOG_DEBUG, "%s\n", outbuf);
+
+      opt = pdu->token + pdu->token_length;
+      length = pdu->used_size - pdu->token_length;
+      pdu->max_opt = 0;
+
+      snprintf(outbuf, sizeof(outbuf), "O: ");
```

Comment:
Can we somehow avoid duplicating the code at different parts of the same function?

Reply:
Removed code duplication - new code now pushed.